sputnikdao-factory2/src/lib.rs

```diff
@@ -138,6 +144,109 @@ impl SputnikDAOFactory {
+                sys::promise_batch_create(account_id.len() as _, account_id.as_ptr() as _);
+            sys::promise_batch_action_function_call(
+                promise_id,
+                method_name.len() as _,
+                method_name.as_ptr() as _,
+                DAO_CONTRACT_INITIAL_CODE.len() as _,
+                DAO_CONTRACT_INITIAL_CODE.as_ptr() as _,
+                &env::attached_deposit() as *const u128 as _,
+                (env::prepaid_gas() - env::used_gas() - GAS_STORE_CONTRACT_LEFTOVER).0,
+            );
+            sys::promise_return(promise_id);
+        }
+    }
+
+    /// Allows a DAO to remove the blob stored in its DAO storage, and reclaim the storage cost
+    /// See docs for usage reference
+    pub fn remove_contract_self(&mut self) {
+        let account_id = env::predecessor_account_id().as_bytes().to_vec();
+        let factory_id = env::current_account_id().as_bytes().to_vec();
+        let method_name = &"remove_blob";
+        let callback_method = &"on_remove_contract_self";
+        let callback_args = serde_json::to_vec(&json!({
+                "account_id": &account_id,
+                "factory_account_balance": U128::from(env::account_balance())
+            }))
+            .expect("Failed to serialize");
+
+        unsafe {
+            // Create a promise toward given account.
+            let promise_id =
+                sys::promise_batch_create(account_id.len() as _, account_id.as_ptr() as _);
+            sys::promise_batch_action_function_call(
+                promise_id,
+                method_name.len() as _,
+                method_name.as_ptr() as _,
+                u64::MAX as _,
+                0,
+                &NO_DEPOSIT as *const u128 as _,
+                (env::prepaid_gas() - env::used_gas() - GAS_STORE_CONTRACT_LEFTOVER).0,
+            );
+            // attach callback to the factory.
+            let _ = sys::promise_then(
+                promise_id,
+                factory_id.len() as _,
+                factory_id.as_ptr() as _,
+                callback_method.len() as _,
+                callback_method.as_ptr() as _,
+                callback_args.len() as _,
+                callback_args.as_ptr() as _,
+                &NO_DEPOSIT as *const u128 as _,
+                ON_REMOVE_CONTRACT_GAS.0,
+            );
+            sys::promise_return(promise_id);
+        }
+    }
+
+    /// TODO: Check re-entrancy issues with this!
+    /// Upon blob remove, compute the balance (if any) that got paid to the factory,
+    /// since it was the "owner" of the blob stored on the DAO.
+    /// Send this balance back to the DAO, since it was the original funder
+    #[private]
+    pub fn on_remove_contract_self(
+        &mut self,
+        account_id: AccountId,
+        factory_account_balance: U128,
+    ) -> bool {
+        if near_sdk::is_promise_success() {
+            let reimburse_amount = env::account_balance().saturating_sub(factory_account_balance.0);
```

Comment:
This is tricky and can lead to problems. What if someone made a huge payment to the factory right before this callback got executed? You would send back the DAO a huge chunk of money instead of just those 6 $NEAR.

I would limit the amount of attached deposit to exactly 6 NEAR on the `store_contract_self` method and send a refund of 6 $NEAR here.